File: ScaleformUI_Csharp/Menus/Pause Menus/LobbyMenu/Items/MissionItem.cs
using ScaleformUI.LobbyMenu;
using ScaleformUI.Menu;
using ScaleformUI.PauseMenu;
using ScaleformUI.Scaleforms;

namespace ScaleformUI
{
    public class MissionItem
    {
        private bool enabled = true;

        public MissionsListColumn ParentColumn { get; internal set; }
        public string Label { get; private set; }
        public HudColor MainColor { get; private set; } = HudColor.HUD_COLOUR_PAUSE_BG;
        public HudColor HighlightColor { get; private set; } = HudColor.HUD_COLOUR_WHITE;
        public BadgeIcon LeftIcon { get; private set; } = BadgeIcon.NONE;
        public HudColor LeftIconColor { get; private set; } = HudColor.HUD_COLOUR_WHITE;
        public BadgeIcon RightIcon { get; private set; } = BadgeIcon.NONE;
        public HudColor RightIconColor { get; private set; } = HudColor.HUD_COLOUR_WHITE;
        public bool RightIconChecked { get; private set; }
        public bool Selected { get; internal set; }
        public bool Enabled
        {
            get => enabled; set
            {
                enabled = value;
                if (ParentColumn != null)
                {
                    if (ParentColumn.Parent is MainView lobby)
                    {
                        lobby._pause._lobby.CallFunction("SET_MISSION_ITEM_ENABLED", ParentColumn.Items.IndexOf(this), value);
                    }
                    else if (ParentColumn.Parent is TabView pause)
                    {
                        pause._pause._pause.CallFunction("SET_PLAYERS_TAB_MISSION_ITEM_ENABLED", ParentColumn.Items.IndexOf(this), value);
                    }
                }
            }
        }

        public MissionItem(string label) : this(label, HudColor.HUD_COLOUR_PAUSE_BG, HudColor.HUD_COLOUR_WHITE) { }

        public MissionItem(string label, HudColor mainColor, HudColor highlightColor)
        {
            Label = label;
            MainColor = mainColor;
            HighlightColor = highlightColor;
        }

        public void SetLeftIcon(BadgeIcon icon, HudColor color)
        {
            LeftIcon = icon;
            LeftIconColor = color;
            if (ParentColumn != null)
            {
                if (ParentColumn.Parent is MainView lobby)
                {
                    lobby._pause._lobby.CallFunction("SET_MISSION_ITEM_LEFT_ICON", (int)icon, color);
                }
                else if (ParentColumn.Parent is TabView pause)
                {
                    pause._pause._pause.CallFunction("SET_PLAYERSTAB_MISSION_ITEM_LEFT_ICON", ParentColumn.ParentTab, (int)icon, color);
                }
            }
        }
        public void SetRightIcon(BadgeIcon icon, HudColor color, bool @checked = false)
        {
            RightIcon = icon;
            RightIconColor = color;
            RightIconChecked = @checked;
            if (ParentColumn != null)
            {
                if (ParentColumn.Parent is MainView lobby)
                {
                    lobby._pause._lobby.CallFunction("SET_MISSION_ITEM_RIGHT_ICON", (int)icon, @checked, color);
                }
                else if (ParentColumn.Parent is TabView pause)
                {
                    pause._pause._pause.CallFunction("SET_PLAYERSTAB_MISSION_ITEM_RIGHT_ICON", ParentColumn.ParentTab, (int)icon, @checked, color);
                }
            }
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/LobbyMenu/Items/MissionItem.lua
MissionItem = setmetatable({}, MissionItem)
MissionItem.__index = MissionItem
MissionItem.__call = function()
    return "LobbyItem", "MissionItem"
end

---@param label string
---@param mainColor SColor
---@param highlightColor SColor
---@return table
function MissionItem.New(label, mainColor, highlightColor)
    local _data = {
        Label = label,
        Handle = nil,
        ParentColumn = nil,
        enabled = true,
        MainColor = mainColor or SColor.HUD_Pause_bg,
        HighlightColor = highlightColor or SColor.HUD_White,
        LeftIcon = BadgeStyle.NONE,
        LeftIconColor = SColor.HUD_White,
        RightIcon = BadgeStyle.NONE,
        RightIconColor = SColor.HUD_White,
        RightIconChecked = false,
        _Selected = false,
        hovered = false,
        Activated = function(item)
        end
    }
    return setmetatable(_data, MissionItem)
end

function MissionItem:Hovered(val)
    if val == nil then
        return self.hovered
    else
        self.hovered = val
    end
end

function MissionItem:Enabled(bool)
    if bool == nil then
        return self.enabled
    else
        self.enabled = bool
        if self.ParentColumn ~= nil and self.ParentColumn.Parent ~= nil and self.ParentColumn.Parent:Visible() then
            local idx = self.ParentColumn.Pagination:GetScaleformIndex(IndexOf(self.ParentColumn.Items, self))
            local pSubT = self.ParentColumn.Parent()
            if pSubT == "LobbyMenu" then
                ScaleformUI.Scaleforms._pauseMenu._lobby:CallFunction("SET_MISSION_ITEM_ENABLED", idx, bool)
            elseif pSubT == "PauseMenu" then
                ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_PLAYERS_TAB_MISSION_ITEM_ENABLED", self.ParentColumn.ParentTab, idx, bool)
            end
        end
    end
end

function MissionItem:SetLeftIcon(icon, color)
    self.LeftIcon = icon
    self.LeftIconColor = color
    if self.ParentColumn ~= nil and self.ParentColumn.Parent ~= nil and self.ParentColumn.Parent:Visible() then
        local idx = self.ParentColumn.Pagination:GetScaleformIndex(IndexOf(self.ParentColumn.Items, self))
        local pSubT = self.ParentColumn.Parent()
        if pSubT == "LobbyMenu" then
            ScaleformUI.Scaleforms._pauseMenu._lobby:CallFunction("SET_MISSION_ITEM_LEFT_ICON", idx, icon, color)
        elseif pSubT == "PauseMenu" then
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_PLAYERS_TAB_MISSION_ITEM_LEFT_ICON", self.ParentColumn.ParentTab, idx, icon, color)
        end
    end
end

function MissionItem:SetRightIcon(icon, color, checked)
    self.RightIcon = icon
    self.RightIconColor = color
    self.RightIconChecked = checked or false
    if self.ParentColumn ~= nil and self.ParentColumn.Parent ~= nil and self.ParentColumn.Parent:Visible() then
        local idx = self.ParentColumn.Pagination:GetScaleformIndex(IndexOf(self.ParentColumn.Items, self))
        local pSubT = self.ParentColumn.Parent()
        if pSubT == "LobbyMenu" then
            ScaleformUI.Scaleforms._pauseMenu._lobby:CallFunction("SET_MISSION_ITEM_RIGHT_ICON", idx, icon, checked, color)
        elseif pSubT == "PauseMenu" then
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_PLAYERS_TAB_MISSION_ITEM_RIGHT_ICON", self.ParentColumn.ParentTab, idx, icon, checked or false, color)
        end
    end
end

function MissionItem:Selected(bool)
    if bool == nil then
        return self._Selected
    else
        self._Selected = bool
    end
end